File: ecosystem/agent-runtime/loom/web/loom-web/src/lib/crash/self-monitoring.ts
/**
 * Copyright (c) 2025 Geoffrey Huntley <ghuntley@ghuntley.com>. All rights reserved.
 * SPDX-License-Identifier: Proprietary
 */

/**
 * Self-monitoring for loom-web.
 *
 * Automatically initializes crash monitoring for the web application by
 * fetching configuration from the server. This enables Loom to monitor itself
 * without any user configuration.
 */

import { CrashClient } from '@loom/crash';
import { browser } from '$app/environment';

/**
 * Configuration returned by the self-monitoring API.
 */
interface WebCrashConfig {
	project_id: string;
	api_key: string;
	release: string;
	environment: string;
}

/**
 * Global crash client instance.
 */
let crashClient: CrashClient | null = null;

/**
 * Initialize self-monitoring for loom-web.
 *
 * This fetches the crash SDK configuration from the server and initializes
 * the CrashClient with global error handlers.
 *
 * Call this once during application initialization.
 *
 * @param debug - Enable debug logging
 * @returns The CrashClient instance, or null if initialization failed
 */
export async function initializeSelfMonitoring(
	debug = false
): Promise<CrashClient | null> {
	// Only run in browser
	if (!browser) {
		return null;
	}

	// Don't initialize twice
	if (crashClient) {
		return crashClient;
	}

	try {
		// Fetch configuration from server
		const response = await fetch('/api/self-monitoring/web-config');

		if (!response.ok) {
			if (debug) {
				console.warn(
					'[Crash] Self-monitoring not available:',
					response.status,
					response.statusText
				);
			}
			return null;
		}

		const config: WebCrashConfig = await response.json();

		// Create crash client
		crashClient = new CrashClient({
			baseUrl: window.location.origin,
			project: config.project_id,
			apiKey: config.api_key,
			release: config.release,
			environment: config.environment,
			debug,
			sessionTracking: true
		});

		// Install global error handlers
		crashClient.installGlobalHandler();

		if (debug) {
			console.log('[Crash] Self-monitoring initialized for loom-web');
		}

		return crashClient;
	} catch (error) {
		if (debug) {
			console.error('[Crash] Failed to initialize self-monitoring:', error);
		}
		return null;
	}
}

/**
 * Get the current crash client instance.
 *
 * Returns null if self-monitoring has not been initialized.
 */
export function getCrashClient(): CrashClient | null {
	return crashClient;
}

/**
 * Capture an exception using the self-monitoring crash client.
 *
 * This is a convenience function that handles the case where
 * self-monitoring is not initialized.
 *
 * @param error - The error to capture
 * @returns The event ID, or empty string if not initialized
 */
export function captureException(error: unknown): string {
	if (!crashClient) {
		return '';
	}
	return crashClient.captureException(error);
}

/**
 * Capture a message using the self-monitoring crash client.
 *
 * @param message - The message to capture
 * @returns The event ID, or empty string if not initialized
 */
export function captureMessage(message: string): string {
	if (!crashClient) {
		return '';
	}
	return crashClient.captureMessage(message);
}

/**
 * Shutdown the self-monitoring crash client.
 *
 * Call this during application shutdown to flush pending events.
 */
export async function shutdownSelfMonitoring(): Promise<void> {
	if (crashClient) {
		await crashClient.shutdown();
		crashClient = null;
	}
}

File: ecosystem/agent-runtime/loom/crates/loom-cli/src/self_monitoring.rs
// Copyright (c) 2025 Geoffrey Huntley <ghuntley@ghuntley.com>. All rights reserved.
// SPDX-License-Identifier: Proprietary

//! Self-monitoring for loom-cli.
//!
//! Automatically initializes crash monitoring for the CLI by fetching
//! configuration from the server. This enables Loom to monitor itself
//! without any user configuration.

use std::sync::Arc;

use anyhow::Result;
use loom_crash::{CrashClient, CrashClientBuilder};
use serde::Deserialize;
use tokio::sync::OnceCell;
use tracing::{debug, warn};

/// Well-known project ID for loom-cli (from loom-server self_monitoring.rs).
const LOOM_CLI_PROJECT_ID: &str = "00000000-0000-0000-0000-000000000004";

/// Global crash client for loom-cli.
static CLI_CRASH_CLIENT: OnceCell<Arc<CrashClient>> = OnceCell::const_new();

/// Configuration returned by the self-monitoring API.
#[derive(Debug, Deserialize)]
struct CliCrashConfig {
	project_id: String,
	api_key: String,
	release: String,
	environment: String,
}

/// Initialize self-monitoring for loom-cli.
///
/// This fetches the CLI API key from the server's self-monitoring endpoint
/// and initializes the crash client with a panic hook.
///
/// Call this once during CLI startup. Failures are logged but do not
/// prevent the CLI from running.
pub async fn initialize_self_monitoring(server_url: &str) -> Option<Arc<CrashClient>> {
	// Don't initialize twice
	if let Some(client) = CLI_CRASH_CLIENT.get() {
		return Some(client.clone());
	}

	match fetch_and_initialize(server_url).await {
		Ok(client) => {
			let client = Arc::new(client);
			let _ = CLI_CRASH_CLIENT.set(client.clone());
			debug!("Self-monitoring initialized for loom-cli");
			Some(client)
		}
		Err(e) => {
			warn!("Failed to initialize self-monitoring: {}", e);
			None
		}
	}
}

async fn fetch_and_initialize(server_url: &str) -> Result<CrashClient> {
	// Fetch CLI config from server
	let http = loom_common_http::new_client();
	let url = format!(
		"{}/api/self-monitoring/cli-config",
		server_url.trim_end_matches('/')
	);

	let response = http.get(&url).send().await?;

	if !response.status().is_success() {
		anyhow::bail!(
			"Self-monitoring not available: {} {}",
			response.status(),
			response.status().canonical_reason().unwrap_or("")
		);
	}

	let config: CliCrashConfig = response.json().await?;

	let client = CrashClientBuilder::new()
		.base_url(server_url)
		.project_id(&config.project_id)
		.auth_token(&config.api_key)
		.release(&config.release)
		.environment(&config.environment)
		.build()?;

	// Install panic hook for automatic crash reporting
	client.install_panic_hook();

	Ok(client)
}

/// Get the CLI crash client for manual error capture.
pub fn get_crash_client() -> Option<Arc<CrashClient>> {
	CLI_CRASH_CLIENT.get().cloned()
}

/// Capture an exception using the self-monitoring crash client.
///
/// This is a convenience function that handles the case where
/// self-monitoring is not initialized.
#[allow(dead_code)]
pub fn capture_exception(error: &dyn std::error::Error) -> Option<String> {
	if let Some(client) = get_crash_client() {
		// Use blocking capture in sync context
		// For now, just log since async capture would require runtime
		warn!("CLI crash captured (async send not implemented): {}", error);
		return None;
	}
	None
}

/// Shutdown the CLI crash client.
///
/// Call this during CLI shutdown to flush pending events.
pub async fn shutdown_self_monitoring() {
	if let Some(client) = CLI_CRASH_CLIENT.get() {
		if let Err(e) = client.shutdown().await {
			warn!("Failed to shutdown crash client: {}", e);
		}
	}
}
